File: examples/networking/tcp_mon_block/src/tcp_mon_block.c
/*author: https://github.com/agentzex
Licensed under the Apache License, Version 2.0 (the "License")

tcp_mon_block.c - uses netlink TC, kernel tracepoints and kprobes to monitor outgoing connections from given PIDs
and block connections to all addresses initiated from them (acting like an in-process firewall), unless they are listed in allow_list
*/

#include <uapi/linux/bpf.h>
#include <uapi/linux/ptrace.h>
#include <linux/tcp.h>
#include <net/sock.h>
#include <net/inet_sock.h>
#include <linux/in.h>
#include <linux/if_ether.h>
#include <linux/if_packet.h>
#include <uapi/linux/if_ether.h>
#include <uapi/linux/ip.h>
#include <linux/tcp.h>
#include <uapi/linux/pkt_cls.h>
#include <linux/fs.h>
#include <linux/uaccess.h>


typedef struct
{
    u32 src_ip;
    u16 src_port;
    u32 dst_ip;
    u16 dst_port;
    u32 pid;
    u8 tcp_flags;
    char comm[TASK_COMM_LEN];
} full_packet;


typedef struct
{
    u8 state;
    u32 src_ip;
    u16 src_port;
    u32 dst_ip;
    u16 dst_port;
    u32 pid;
    char comm[TASK_COMM_LEN];
} verbose_event;


typedef struct
{
    u32 src_ip;
    u16 src_port;
    u32 dst_ip;
    u16 dst_port;
} key_hash;


BPF_HASH(monitored_connections, key_hash, full_packet);
BPF_HASH(allow_list, u32, u32);
BPF_HASH(pid_list, u32, u32);
BPF_PERF_OUTPUT(blocked_events);
BPF_PERF_OUTPUT(verbose_events);


#ifndef tcp_flag_byte
#define tcp_flag_byte(th) (((u_int8_t *)th)[13])
#endif


static bool VERBOSE_OUTPUT = false;


static __always_inline int tcp_header_bound_check(struct tcphdr* tcp, void* data_end)
{
    if ((void *)tcp + sizeof(*tcp) > data_end)
    {
        return -1;
    }

    return 0;
}


static void make_verbose_event(verbose_event *v, u32 src_ip, u32 dst_ip, u16 src_port, u16 dst_port, u32 pid, u8 state)
{
    v->src_ip = src_ip;
    v->src_port = src_port;
    v->dst_ip = dst_ip;
    v->dst_port = dst_port;
    v->pid = pid;
    v->state = state;
    bpf_get_current_comm(&v->comm, sizeof(v->comm));
}


int handle_egress(struct __sk_buff *ctx)
{
    void* data_end = (void*)(long)ctx->data_end;
    void* data = (void*)(long)ctx->data;
    struct ethhdr *eth = data;
    struct iphdr *ip = data + sizeof(*eth);
    struct tcphdr *tcp;
    key_hash key = {};

    /* length check */
    if (data + sizeof(*eth) + sizeof(*ip) > data_end)
    {
        return TC_ACT_OK;
    }

    if (eth->h_proto != htons(ETH_P_IP))
    {
        return TC_ACT_OK;
    }

    if (ip->protocol != IPPROTO_TCP)
    {
        return TC_ACT_OK;
    }

    tcp = (void *)ip + sizeof(*ip);
    if (tcp_header_bound_check(tcp, data_end))
    {
        return TC_ACT_OK;
    }

    u8 tcpflags = ((u_int8_t *)tcp)[13];
    u16 src_port = bpf_ntohs(tcp->source);
    u16 dst_port = bpf_ntohs(tcp->dest);

    key.src_ip = ip->saddr;
    key.src_port = src_port;
    key.dst_ip = ip->daddr;
    key.dst_port = dst_port;

    full_packet *packet_value;
    packet_value = monitored_connections.lookup(&key);
    if (packet_value != 0)
    {
        packet_value->tcp_flags = tcpflags;
        blocked_events.perf_submit(ctx, packet_value, sizeof(full_packet));
        return TC_ACT_SHOT;
    }

    return TC_ACT_OK;
}


// Removing the entry from monitored_connections when the socket closes after failed connection
TRACEPOINT_PROBE(sock, inet_sock_set_state)
{
    if (args->protocol != IPPROTO_TCP)
    {
        return 0;
    }

    if (args->newstate != TCP_CLOSE && args->newstate != TCP_CLOSE_WAIT)
    {
        return 0;
    }

    if (args->family == AF_INET)
    {
        key_hash key = {};
        struct sock *sk = (struct sock *)args->skaddr;

        key.src_port = args->sport;
        key.dst_port = args->dport;
        __builtin_memcpy(&key.src_ip, args->saddr, sizeof(key.src_ip));
        __builtin_memcpy(&key.dst_ip, args->daddr, sizeof(key.dst_ip));

        full_packet *packet_value;
        packet_value = monitored_connections.lookup(&key);
        if (packet_value != 0)
        {
            monitored_connections.delete(&key);
            if (VERBOSE_OUTPUT)
            {
                verbose_event v = {};
                make_verbose_event(&v, packet_value->src_ip, packet_value->dst_ip, packet_value->src_port, packet_value->dst_port, packet_value->pid, 3);
                verbose_events.perf_submit(args, &v, sizeof(v));
            }

        }
    }

    return 0;
}




int trace_connect_entry(struct pt_regs *ctx, struct sock *sk)
{
    key_hash key = {};
    full_packet packet_value = {};
    u8 verbose_state = 0;

    u16 family = sk->__sk_common.skc_family;
    if (family != AF_INET)
    {
        return 0;
    }

    u32 pid = bpf_get_current_pid_tgid() >> 32;
    u16 dst_port = sk->__sk_common.skc_dport;
    dst_port = ntohs(dst_port);
    u16 src_port = sk->__sk_common.skc_num;
    u32 src_ip = sk->__sk_common.skc_rcv_saddr;
    u32 dst_ip = sk->__sk_common.skc_daddr;

    u32 *monitored_pid = pid_list.lookup(&pid);
    if (!monitored_pid)
    {
        return 0;
    }

    u32 *allowed_ip = allow_list.lookup(&dst_ip);
    if (!allowed_ip)
    {
        key.src_ip = src_ip;
        key.src_port = src_port;
        key.dst_ip = dst_ip;
        key.dst_port = dst_port;

        packet_value.src_ip = src_ip;
        packet_value.src_port = src_port;
        packet_value.dst_ip = dst_ip;
        packet_value.dst_port = dst_port;
        packet_value.pid = pid;
        bpf_get_current_comm(&packet_value.comm, sizeof(packet_value.comm));
        verbose_state = 1;
        monitored_connections.update(&key, &packet_value);
    }
    else
    {
        verbose_state = 2;
    }

    if (VERBOSE_OUTPUT)
    {
        verbose_event v = {};
        make_verbose_event(&v, src_ip, dst_ip, src_port, dst_port, pid, verbose_state);
        verbose_events.perf_submit(ctx, &v, sizeof(v));
    }

    return 0;
}

File: examples/networking/tcp_mon_block/src/tcp_mon_block.py
#!/usr/bin/python
# author: https://github.com/agentzex
# Licensed under the Apache License, Version 2.0 (the "License")

# tcp_mon_block.py - uses netlink TC, kernel tracepoints and kprobes to monitor outgoing connections from given PIDs
# and block connections to all addresses initiated from them (acting like an in-process firewall), unless they are listed in allow_list

# outputs blocked connections attempts from monitored processes
# Usage:
#   python3 tcp_mon_block.py -i network_interface_name
#   python3 tcp_mon_block.py -v -i network_interface_name (-v --verbose - will output all connections attempts, including allowed ones)
#


from bcc import BPF
import pyroute2
import socket
import struct
import json
import argparse
from urllib.parse import urlparse


# TCP flags
FIN = 0x01
SYN = 0x02
RST = 0x04
PSH = 0x08
ACK = 0x10
URG = 0x20
ECE = 0x40
CWR = 0x80


verbose_states = {
    1: "Connection not allowed detected - forwarding to block",
    2: "Connection allowed",
    3: "Connection destroyed",
}


def get_verbose_message(state):
    if state not in verbose_states:
        return ""

    return verbose_states[state]


def parse_tcp_flags(flags):
    found_flags = ""
    if flags & FIN:
        found_flags += "FIN; "
    if flags & SYN:
        found_flags += "SYN; "
    if flags & RST:
        found_flags += "RST; "
    if flags & PSH:
        found_flags += "PSH; "
    if flags & ACK:
        found_flags += "ACK; "
    if flags & URG:
        found_flags += "URG; "
    if flags & ECE:
        found_flags += "ECE; "
    if flags & CWR:
        found_flags += "CWR;"

    return found_flags


def ip_to_network_address(ip):
    return struct.unpack("I", socket.inet_aton(ip))[0]


def network_address_to_ip(ip):
    return socket.inet_ntop(socket.AF_INET, struct.pack("I", ip))


def parse_address(url_or_ip):
    is_ipv4 = True
    domain = ""

    #first check if valid ipv4
    try:
        socket.inet_aton(url_or_ip)
    except socket.error:
        is_ipv4 = False

    if is_ipv4:
        return [url_or_ip]

    # if not check if valid URL, parse and get its domain, resolve it to IPv4 and return it
    try:
        domain = urlparse(url_or_ip).netloc
    except:
        print(f"[-] {url_or_ip} is invalid IPv4 or URL")
        return False

    # should get a list of IPv4 addresses resolved from the domain
    try:
        hostname, aliaslist, ipaddrlist = socket.gethostbyname_ex(domain)
    except:
        print(f"[-] Failed to resolve {url_or_ip} to Ipv4")
        return False

    return ipaddrlist


def create_bpf_allow_list(bpf):
    bpf_allow_list = bpf.get_table("allow_list")
    bpf_pid_list = bpf.get_table("pid_list")
    with open("allow_list.json", "r") as f:
        pids_to_list = json.loads(f.read())

    print("[+] Reading and parsing allow_list.json")
    for pid_to_list in pids_to_list:
        try:
            pid = int(pid_to_list["pid"])
        except ValueError:
            print(f"[-] invalid PID: {pid_to_list['pid']}")
            continue

        print(f"[+] Adding {pid} to monitored processes")
        bpf_pid_list[bpf_pid_list.Key(pid)] = bpf_pid_list.Leaf(pid)

        for url_or_ip in pid_to_list["allow_list"]:
            ips = parse_address(url_or_ip)
            if not ips:
                continue
            for ip in ips:
                print(f"[+] Adding {ip} to allowed IPs")
                ip = ip_to_network_address(ip)
                bpf_allow_list[bpf_allow_list.Key(ip)] = bpf_allow_list.Leaf(ip)


def create_tc(interface):
    ip = pyroute2.IPRoute()
    ipdb = pyroute2.IPDB(nl=ip)
    try:
        idx = ipdb.interfaces[interface].index
    except:
        print(f"[-] {interface} interface not found")
        return False, False, False

    try:
        # deleting if exists from previous run
        ip.tc("del", "clsact", idx)
    except:
        pass
    ip.tc("add", "clsact", idx)
    return ip, ipdb, idx


def parse_blocked_event(cpu, data, size):
    event = bpf["blocked_events"].event(data)
    src_ip = network_address_to_ip(event.src_ip)
    dst_ip = network_address_to_ip(event.dst_ip)
    flags = parse_tcp_flags(event.tcp_flags)
    print(f"{event.pid}: {event.comm.decode()} - {src_ip}:{event.src_port} -> {dst_ip}:{event.dst_port} Flags: {flags} was blocked!")


def parse_verbose_event(cpu, data, size):
    event = bpf["verbose_events"].event(data)
    src_ip = network_address_to_ip(event.src_ip)
    dst_ip = network_address_to_ip(event.dst_ip)
    verbose_message = get_verbose_message(event.state)
    print(f"{event.pid}: {event.comm.decode()} - {src_ip}:{event.src_port} -> {dst_ip}:{event.dst_port} - {verbose_message}")



parser = argparse.ArgumentParser(description="Monitor given PIDs and block outgoing connections to all addresses initiated from them, unless they are listed in allow_list.json")
parser.add_argument("-i", "--interface", help="Network interface name to monitor traffic on", required=True, type=str)
parser.add_argument("-v", "--verbose", action="store_true", help="Set verbose output")
args = parser.parse_args()
print(f"[+] Monitoring {args.interface} interface")


with open("tcp_mon_block.c", "r") as f:
    bpf_text = f.read()

if args.verbose:
    print("[+] Verbose output is ON!")
    bpf_text = bpf_text.replace("static bool VERBOSE_OUTPUT = false", "static bool VERBOSE_OUTPUT = true")


ip, ipdb, idx = create_tc(args.interface)
if not ip:
    exit(-1)

bpf = BPF(text=bpf_text)
create_bpf_allow_list(bpf)

# loading kprobe
bpf.attach_kprobe(event="tcp_connect", fn_name="trace_connect_entry")

# loading TC
fn = bpf.load_func("handle_egress", BPF.SCHED_CLS)

#default parent handlers:
#https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/uapi/linux/pkt_sched.h?id=1f211a1b929c804100e138c5d3d656992cfd5622
#define TC_H_MIN_INGRESS	0xFFF2U
#define TC_H_MIN_EGRESS		0xFFF3U

ip.tc("add-filter", "bpf", idx, ":1", fd=fn.fd, name=fn.name, parent="ffff:fff3", classid=1, direct_action=True)
bpf["blocked_events"].open_perf_buffer(parse_blocked_event)
bpf["verbose_events"].open_perf_buffer(parse_verbose_event)


print("[+] Monitoring started\n")
while True:
    try:
        bpf.perf_buffer_poll()
    except KeyboardInterrupt:
        break

ip.tc("del", "clsact", idx)
ipdb.release()










